### DISCUSSION TITLE
The cost of change is too high

### DISCUSSION THREAD

**Discussion by tylerjw** [moveit/moveit2#1745] [General]
Currently, many PRs are being abandoned in MoveIt because after the author does work on them, we realize that we disagree with the approach being taken. I want to discuss structural changes we could make to facilitate discussions around the approach taken to changes earlier.

I propose this general structure:

- Discussions (what you are looking at now) is for discussion needs and how various approaches (designs) might meet the need.
- Issues are for iterating over a specific design for solving a problem.
- PRs are for iterating over a specific implementation of a design.

When people skip these steps and submit PRs, there is an opportunity for waste and hurt feelings as PRs go stale, and the need that resulted in someone doing the work on the PR goes unaddressed.

To address that problem, we (as maintainers and mentors for this project) should engage people who want to change MoveIt by discussing their needs and approaches in discussions here.

As we do not have a ton of experience doing this well, I will not propose any specific requirements or processes and instead provide these guidelines:

1. A shared understanding of the need is a prerequisite to discussing designs.
2. An approach should be agreed upon between the one making the change and the maintainers before beginning work on implementation.
3. With a shared understanding of the need and approach, reviews should be inexpensive (in time) for reviewers and result in the experience of making a change of MoveIt more successful.

**Comment by AndyZe** [comment]
> Issues are for iterating over a specific design for solving a problem.

I think Issues should be for quickly flagging bugs or making suggestions. That's kinda what the name implies and it's how people normally use the issue tracker.

Agree that it would be nice sometimes to get more discussion before a PR is made.

**Comment by tylerjw** [comment]
In many ways, @AndyZe, you do a really good job opening PRs that do not require architecture-level discussions most of the time. I want to find ways to make it easier to make significant changes to MoveIt by having discussions about the change before people start work.
